Duplicati/GUI/Duplicati.GUI.TrayIcon/Program.cs

```diff
@@ -133,73 +137,256 @@ public static void RealMain(string[] _args)
-                        Action shutdownEvent = () =>
-                        {
-                            tk.InvokeExit();
-                        };
+            //Create the connection instance
+            var con = Library.SQLiteHelper.SQLiteLoader.LoadConnection();
+
+            try
+            {
+                var databasePath = System.IO.Path.Combine(Library.Utility.Utility.AppendDirSeparator(
+                    Library.Utility.Utility.ExpandEnvironmentVariables(serverDataFolder).Trim('"')), "Duplicati-server.sqlite");
+
+                if (!System.IO.Directory.Exists(System.IO.Path.GetDirectoryName(databasePath)))
+                    System.IO.Directory.CreateDirectory(System.IO.Path.GetDirectoryName(databasePath));
+#if DEBUG
+                //Default is to not use encryption for debugging
+                var useDatabaseEncryption = commandlineOptions.ContainsKey("unencrypted-database") && !Library.Utility.Utility.ParseBool(commandlineOptions["unencrypted-database"], true);
+#else
+                var useDatabaseEncryption = !commandlineOptions.ContainsKey("unencrypted-database") || !Library.Utility.Utility.ParseBool(commandlineOptions["unencrypted-database"], true);
+#endif
+                con.ConnectionString = "Data Source=" + databasePath;
 
-                        if (hosted != null)
-                            hosted.InstanceShutdown += shutdownEvent;
+                //Attempt to open the database, handling any encryption present
+                OpenDatabase(con, useDatabaseEncryption, dbPassword);
 
-                        tk.Init(_args);
+                Duplicati.Library.SQLiteHelper.DatabaseUpgrader.UpgradeDatabase(con, databasePath,
+                    typeof(Duplicati.Server.Database.Connection));
+            }
+            catch (Exception ex)
+            {
+                //Unwrap the reflection exceptions
+                if (ex is System.Reflection.TargetInvocationException && ex.InnerException != null)
+                    ex = ex.InnerException;
+
+                throw new Exception($@"Failed to create, open or upgrade the database. Error message: {ex}");
+            }
+
+            return new Duplicati.Server.Database.Connection(con);
+        }
 
-                        // Make sure that the server shutdown does not access the tray-icon,
-                        // as it would be disposed by now
-                        if (hosted != null)
-                            hosted.InstanceShutdown -= shutdownEvent;
+        /// <summary>
+        /// Helper method with logic to handle opening a database in possibly encrypted format
+        /// </summary>
+        /// <param name="con">The SQLite connection object</param>
+        /// <param name="useDatabaseEncryption">Specify if database is encrypted</param>
+        /// <param name="password">Encryption password</param>
+        private static void OpenDatabase(System.Data.IDbConnection con, bool useDatabaseEncryption, string password)

```

Comment:
Why not make `OpenDatabase` public in `Duplicati.Server` and then call it there?